runtime/ftplugin/javascript.vim

```diff
@@ -70,8 +70,23 @@ let &l:define =
             \ .. '\|^\s*\(export\s\+\|export\s\+default\s\+\)*\(var\|let\|const\|function\|class\)'
             \ .. '\|\<as\>'
 
+if !empty(get(g:,'javascript_formatprg', ''))
+augroup formatprgsJavaScript
+  autocmd! * <buffer>
+  autocmd ShellFilterPost <buffer> if v:shell_error | execute 'echom "shell filter returned error " . v:shell_error . ", undoing changes"' | undo | endif
+  if g:javascript_formatprg ==# 'prettier'
+    autocmd BufWinEnter <buffer> ++once let &l:formatprg =
+              \ executable('prettier') ? 'prettier' : 'npx prettier' .
```

Comment:
Doesn't `npx` install software from the internet? That can be a big deal in some environments, and I really don't think it's a good idea to do that without the user explicitly choosing to do it.

Maybe `g:javascript_formatter` could support a separate `npx-prettier` for users who want that behavior?

Reply:
It's also a bit strange. you are setting it to `npx prettier` although `g:javascript_formatprg` equals `prettier`. That looks a bit strange. Also you seem to assume that all formatters understand the same arguments.

Reply:
@dkearns please take a look at https://github.com/vim/vim/pull/16989#:~:text=3%20days%20ago-,%5BWIP%5D%20Add%20%22formatter%22%20feature%20using%20%22compiler%22%20as%20a%20template%20%2317145,-Draft, which is another approach to the problem discussed here.

> I think we need a global `npx` opt-in. There's also several compilers that use it, or could do so.

Why not? How would you name that variable?

Reply:
> I think we need a global `npx` opt-in. There's also several compilers that use it, or could do so.

I don't have much of an opinion either way. But if it is global, it might be worth thinking about other similar things like `pipx`. E.g., maybe `g:automatically_install_from = ['npx']` would make more sense than `g:automatically_install_from_npx = v:true`. Or maybe not. Just figured I'd bring it up in case it affects the direction of this.

> @dkearns please take a look at #16989, which is another approach to the problem discussed here.

I'm guessing you meant #17145?

Reply:
Sigh… yes, #17145.